File: services/src/main/java/com/liferay/mobile/sdk/v7/ddmtemplate/DDMTemplateService.java
package com.liferay.mobile.sdk.v7.ddmtemplate;

import com.liferay.mobile.sdk.Call;
import com.liferay.mobile.sdk.annotation.Param;
import com.liferay.mobile.sdk.annotation.Path;
import com.liferay.mobile.sdk.file.UploadData;
import com.liferay.mobile.sdk.http.ContentType;
import com.liferay.mobile.sdk.http.Response;
import java.lang.Integer;
import java.lang.String;
import org.json.JSONArray;
import org.json.JSONObject;

@Path("/ddm.ddmtemplate")
public interface DDMTemplateService {
  @Path("/add-template")
  Call<JSONObject> addTemplate(@Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "classPK") long classPK, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "nameMap", className = "") JSONObject nameMap, @Param(name = "descriptionMap", className = "") JSONObject descriptionMap, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "language") String language, @Param(name = "script") String script, @Param(name = "serviceContext", className = "") JSONObject serviceContext);

  @Path(
      value = "/add-template",
      contentType = ContentType.MULTIPART
  )
  Call<JSONObject> addTemplate(@Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "classPK") long classPK, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "templateKey") String templateKey, @Param(name = "nameMap", className = "") JSONObject nameMap, @Param(name = "descriptionMap", className = "") JSONObject descriptionMap, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "language") String language, @Param(name = "script") String script, @Param(name = "cacheable") boolean cacheable, @Param(name = "smallImage") boolean smallImage, @Param(name = "smallImageURL") String smallImageURL, @Param(name = "smallImageFile") UploadData smallImageFile, @Param(name = "serviceContext", className = "") JSONObject serviceContext);

  @Path("/copy-template")
  Call<JSONObject> copyTemplate(@Param(name = "templateId") long templateId, @Param(name = "serviceContext", className = "") JSONObject serviceContext);

  @Path("/copy-template")
  Call<JSONObject> copyTemplate(@Param(name = "templateId") long templateId, @Param(name = "nameMap", className = "") JSONObject nameMap, @Param(name = "descriptionMap", className = "") JSONObject descriptionMap, @Param(name = "serviceContext", className = "") JSONObject serviceContext);

  @Path("/copy-templates")
  Call<JSONArray> copyTemplates(@Param(name = "classNameId") long classNameId, @Param(name = "oldClassPK") long oldClassPK, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "newClassPK") long newClassPK, @Param(name = "type") String type, @Param(name = "serviceContext", className = "") JSONObject serviceContext);

  @Path("/delete-template")
  Call<Response> deleteTemplate(@Param(name = "templateId") long templateId);

  @Path("/fetch-template")
  Call<JSONObject> fetchTemplate(@Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "templateKey") String templateKey);

  @Path("/get-template")
  Call<JSONObject> getTemplate(@Param(name = "templateId") long templateId);

  @Path("/get-template")
  Call<JSONObject> getTemplate(@Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "templateKey") String templateKey);

  @Path("/get-template")
  Call<JSONObject> getTemplate(@Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "templateKey") String templateKey, @Param(name = "includeAncestorTemplates") boolean includeAncestorTemplates);

  @Path("/get-templates")
  Call<JSONArray> getTemplates(@Param(name = "companyId") long companyId, @Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "status") int status);

  @Path("/get-templates")
  Call<JSONArray> getTemplates(@Param(name = "companyId") long companyId, @Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "classPK") long classPK, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "status") int status);

  @Path("/get-templates")
  Call<JSONArray> getTemplates(@Param(name = "companyId") long companyId, @Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "classPK") long classPK, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "type") String type, @Param(name = "status") int status);

  @Path("/get-templates")
  Call<JSONArray> getTemplates(@Param(name = "companyId") long companyId, @Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "classPK") long classPK, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "includeAncestorTemplates") boolean includeAncestorTemplates, @Param(name = "status") int status);

  @Path("/get-templates")
  Call<JSONArray> getTemplates(@Param(name = "companyId") long companyId, @Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "classPK") long classPK, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "status") int status);

  @Path("/get-templates-by-class-pk")
  Call<JSONArray> getTemplatesByClassPK(@Param(name = "companyId") long companyId, @Param(name = "groupId") long groupId, @Param(name = "classPK") long classPK, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "status") int status);

  @Path("/get-templates-by-structure-class-name-id")
  Call<JSONArray> getTemplatesByStructureClassNameId(@Param(name = "groupId") long groupId, @Param(name = "structureClassNameId") long structureClassNameId, @Param(name = "status") int status, @Param(name = "start") int start, @Param(name = "end") int end, @Param(name = "orderByComparator", className = "") JSONObject orderByComparator);

  @Path("/get-templates-by-structure-class-name-id-count")
  Call<Integer> getTemplatesByStructureClassNameIdCount(@Param(name = "groupId") long groupId, @Param(name = "structureClassNameId") long structureClassNameId, @Param(name = "status") int status);

  @Path("/revert-template")
  Call<Response> revertTemplate(@Param(name = "templateId") long templateId, @Param(name = "version") String version, @Param(name = "serviceContext", className = "") JSONObject serviceContext);

  @Path("/search")
  Call<JSONArray> search(@Param(name = "companyId") long companyId, @Param(name = "groupIds") JSONArray groupIds, @Param(name = "classNameIds") JSONArray classNameIds, @Param(name = "classPKs") JSONArray classPKs, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "keywords") String keywords, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "status") int status, @Param(name = "start") int start, @Param(name = "end") int end, @Param(name = "orderByComparator", className = "") JSONObject orderByComparator);

  @Path("/search")
  Call<JSONArray> search(@Param(name = "companyId") long companyId, @Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "classPK") long classPK, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "keywords") String keywords, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "status") int status, @Param(name = "start") int start, @Param(name = "end") int end, @Param(name = "orderByComparator", className = "") JSONObject orderByComparator);

  @Path("/search")
  Call<JSONArray> search(@Param(name = "companyId") long companyId, @Param(name = "groupIds") JSONArray groupIds, @Param(name = "classNameIds") JSONArray classNameIds, @Param(name = "classPKs") JSONArray classPKs, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "name") String name, @Param(name = "description") String description, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "language") String language, @Param(name = "status") int status, @Param(name = "andOperator") boolean andOperator, @Param(name = "start") int start, @Param(name = "end") int end, @Param(name = "orderByComparator", className = "") JSONObject orderByComparator);

  @Path("/search")
  Call<JSONArray> search(@Param(name = "companyId") long companyId, @Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "classPK") long classPK, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "name") String name, @Param(name = "description") String description, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "language") String language, @Param(name = "status") int status, @Param(name = "andOperator") boolean andOperator, @Param(name = "start") int start, @Param(name = "end") int end, @Param(name = "orderByComparator", className = "") JSONObject orderByComparator);

  @Path("/search-count")
  Call<Integer> searchCount(@Param(name = "companyId") long companyId, @Param(name = "groupIds") JSONArray groupIds, @Param(name = "classNameIds") JSONArray classNameIds, @Param(name = "classPKs") JSONArray classPKs, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "keywords") String keywords, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "status") int status);

  @Path("/search-count")
  Call<Integer> searchCount(@Param(name = "companyId") long companyId, @Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "classPK") long classPK, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "keywords") String keywords, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "status") int status);

  @Path("/search-count")
  Call<Integer> searchCount(@Param(name = "companyId") long companyId, @Param(name = "groupIds") JSONArray groupIds, @Param(name = "classNameIds") JSONArray classNameIds, @Param(name = "classPKs") JSONArray classPKs, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "name") String name, @Param(name = "description") String description, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "language") String language, @Param(name = "status") int status, @Param(name = "andOperator") boolean andOperator);

  @Path("/search-count")
  Call<Integer> searchCount(@Param(name = "companyId") long companyId, @Param(name = "groupId") long groupId, @Param(name = "classNameId") long classNameId, @Param(name = "classPK") long classPK, @Param(name = "resourceClassNameId") long resourceClassNameId, @Param(name = "name") String name, @Param(name = "description") String description, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "language") String language, @Param(name = "status") int status, @Param(name = "andOperator") boolean andOperator);

  @Path("/update-template")
  Call<JSONObject> updateTemplate(@Param(name = "templateId") long templateId, @Param(name = "classPK") long classPK, @Param(name = "nameMap", className = "") JSONObject nameMap, @Param(name = "descriptionMap", className = "") JSONObject descriptionMap, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "language") String language, @Param(name = "script") String script, @Param(name = "cacheable") boolean cacheable, @Param(name = "serviceContext", className = "") JSONObject serviceContext);

  @Path(
      value = "/update-template",
      contentType = ContentType.MULTIPART
  )
  Call<JSONObject> updateTemplate(@Param(name = "templateId") long templateId, @Param(name = "classPK") long classPK, @Param(name = "nameMap", className = "") JSONObject nameMap, @Param(name = "descriptionMap", className = "") JSONObject descriptionMap, @Param(name = "type") String type, @Param(name = "mode") String mode, @Param(name = "language") String language, @Param(name = "script") String script, @Param(name = "cacheable") boolean cacheable, @Param(name = "smallImage") boolean smallImage, @Param(name = "smallImageURL") String smallImageURL, @Param(name = "smallImageFile") UploadData smallImageFile, @Param(name = "serviceContext", className = "") JSONObject serviceContext);
}

File: windows/Liferay.SDK/Service/V62/DDMTemplate/DDMTemplateService.cs
/**
 * Copyright (c) 2000-present Liferay, Inc. All rights reserved.
 *
 * This library is free software; you can redistribute it and/or modify it under
 * the terms of the GNU Lesser General Public License as published by the Free
 * Software Foundation; either version 2.1 of the License, or (at your option)
 * any later version.
 *
 * This library is distributed in the hope that it will be useful, but WITHOUT
 * ANY WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS
 * FOR A PARTICULAR PURPOSE. See the GNU Lesser General Public License for more
 * details.
 */

using System;
using System.Collections.Generic;
using System.IO;
using System.Linq;
using System.Threading.Tasks;

namespace Liferay.SDK.Service.V62.DDMTemplate
{
	public class DDMTemplateService : ServiceBase
	{
		public DDMTemplateService(ISession session)
			: base(session)
		{
		}

		public async Task<IDictionary<string, object>> AddTemplateAsync(long groupId, long classNameId, long classPK, IDictionary<string, string> nameMap, IDictionary<string, string> descriptionMap, string type, string mode, string language, string script, IDictionary<string, object> serviceContext)
		{
			var _parameters = new JsonObject();

			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);
			_parameters.Add("classPK", classPK);
			_parameters.Add("nameMap", nameMap);
			_parameters.Add("descriptionMap", descriptionMap);
			_parameters.Add("type", type);
			_parameters.Add("mode", mode);
			_parameters.Add("language", language);
			_parameters.Add("script", script);
			_parameters.Add("serviceContext", serviceContext);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/add-template", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> AddTemplateAsync(long groupId, long classNameId, long classPK, string templateKey, IDictionary<string, string> nameMap, IDictionary<string, string> descriptionMap, string type, string mode, string language, string script, bool cacheable, bool smallImage, string smallImageURL, IDictionary<string, object> smallImageFile, IDictionary<string, object> serviceContext)
		{
			var _parameters = new JsonObject();

			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);
			_parameters.Add("classPK", classPK);
			_parameters.Add("templateKey", templateKey);
			_parameters.Add("nameMap", nameMap);
			_parameters.Add("descriptionMap", descriptionMap);
			_parameters.Add("type", type);
			_parameters.Add("mode", mode);
			_parameters.Add("language", language);
			_parameters.Add("script", script);
			_parameters.Add("cacheable", cacheable);
			_parameters.Add("smallImage", smallImage);
			_parameters.Add("smallImageURL", smallImageURL);
			_parameters.Add("smallImageFile", smallImageFile);
			_parameters.Add("serviceContext", serviceContext);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/add-template", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> CopyTemplateAsync(long templateId, IDictionary<string, object> serviceContext)
		{
			var _parameters = new JsonObject();

			_parameters.Add("templateId", templateId);
			_parameters.Add("serviceContext", serviceContext);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/copy-template", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> CopyTemplateAsync(long templateId, IDictionary<string, string> nameMap, IDictionary<string, string> descriptionMap, IDictionary<string, object> serviceContext)
		{
			var _parameters = new JsonObject();

			_parameters.Add("templateId", templateId);
			_parameters.Add("nameMap", nameMap);
			_parameters.Add("descriptionMap", descriptionMap);
			_parameters.Add("serviceContext", serviceContext);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/copy-template", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> CopyTemplatesAsync(long classNameId, long classPK, long newClassPK, string type, IDictionary<string, object> serviceContext)
		{
			var _parameters = new JsonObject();

			_parameters.Add("classNameId", classNameId);
			_parameters.Add("classPK", classPK);
			_parameters.Add("newClassPK", newClassPK);
			_parameters.Add("type", type);
			_parameters.Add("serviceContext", serviceContext);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/copy-templates", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task DeleteTemplateAsync(long templateId)
		{
			var _parameters = new JsonObject();

			_parameters.Add("templateId", templateId);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/delete-template", _parameters }
			};

			await this.Session.InvokeAsync(_command);
		}

		public async Task<IDictionary<string, object>> FetchTemplateAsync(long groupId, long classNameId, string templateKey)
		{
			var _parameters = new JsonObject();

			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);
			_parameters.Add("templateKey", templateKey);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/fetch-template", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> GetTemplateAsync(long templateId)
		{
			var _parameters = new JsonObject();

			_parameters.Add("templateId", templateId);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/get-template", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> GetTemplateAsync(long groupId, long classNameId, string templateKey)
		{
			var _parameters = new JsonObject();

			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);
			_parameters.Add("templateKey", templateKey);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/get-template", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> GetTemplateAsync(long groupId, long classNameId, string templateKey, bool includeGlobalTemplates)
		{
			var _parameters = new JsonObject();

			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);
			_parameters.Add("templateKey", templateKey);
			_parameters.Add("includeGlobalTemplates", includeGlobalTemplates);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/get-template", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> GetTemplatesAsync(long groupId, long classNameId)
		{
			var _parameters = new JsonObject();

			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/get-templates", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> GetTemplatesAsync(long groupId, long classNameId, long classPK)
		{
			var _parameters = new JsonObject();

			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);
			_parameters.Add("classPK", classPK);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/get-templates", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> GetTemplatesAsync(long groupId, long classNameId, long classPK, string type)
		{
			var _parameters = new JsonObject();

			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);
			_parameters.Add("classPK", classPK);
			_parameters.Add("type", type);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/get-templates", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> GetTemplatesAsync(long groupId, long classNameId, long classPK, string type, string mode)
		{
			var _parameters = new JsonObject();

			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);
			_parameters.Add("classPK", classPK);
			_parameters.Add("type", type);
			_parameters.Add("mode", mode);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/get-templates", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> GetTemplatesByClassPkAsync(long groupId, long classPK)
		{
			var _parameters = new JsonObject();

			_parameters.Add("groupId", groupId);
			_parameters.Add("classPK", classPK);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/get-templates-by-class-pk", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> GetTemplatesByStructureClassNameIdAsync(long groupId, long structureClassNameId, int start, int end, IDictionary<string, object> orderByComparator)
		{
			var _parameters = new JsonObject();

			_parameters.Add("groupId", groupId);
			_parameters.Add("structureClassNameId", structureClassNameId);
			_parameters.Add("start", start);
			_parameters.Add("end", end);
			_parameters.Add("orderByComparator", orderByComparator);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/get-templates-by-structure-class-name-id", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<long> GetTemplatesByStructureClassNameIdCountAsync(long groupId, long structureClassNameId)
		{
			var _parameters = new JsonObject();

			_parameters.Add("groupId", groupId);
			_parameters.Add("structureClassNameId", structureClassNameId);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/get-templates-by-structure-class-name-id-count", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (long)_obj;
		}

		public async Task<IDictionary<string, object>> SearchAsync(long companyId, long groupId, long classNameId, long classPK, string keywords, string type, string mode, int start, int end, IDictionary<string, object> orderByComparator)
		{
			var _parameters = new JsonObject();

			_parameters.Add("companyId", companyId);
			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);
			_parameters.Add("classPK", classPK);
			_parameters.Add("keywords", keywords);
			_parameters.Add("type", type);
			_parameters.Add("mode", mode);
			_parameters.Add("start", start);
			_parameters.Add("end", end);
			_parameters.Add("orderByComparator", orderByComparator);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/search", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> SearchAsync(long companyId, IEnumerable<long> groupIds, IEnumerable<long> classNameIds, IEnumerable<long> classPKs, string keywords, string type, string mode, int start, int end, IDictionary<string, object> orderByComparator)
		{
			var _parameters = new JsonObject();

			_parameters.Add("companyId", companyId);
			_parameters.Add("groupIds", groupIds);
			_parameters.Add("classNameIds", classNameIds);
			_parameters.Add("classPKs", classPKs);
			_parameters.Add("keywords", keywords);
			_parameters.Add("type", type);
			_parameters.Add("mode", mode);
			_parameters.Add("start", start);
			_parameters.Add("end", end);
			_parameters.Add("orderByComparator", orderByComparator);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/search", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> SearchAsync(long companyId, long groupId, long classNameId, long classPK, string name, string description, string type, string mode, string language, bool andOperator, int start, int end, IDictionary<string, object> orderByComparator)
		{
			var _parameters = new JsonObject();

			_parameters.Add("companyId", companyId);
			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);
			_parameters.Add("classPK", classPK);
			_parameters.Add("name", name);
			_parameters.Add("description", description);
			_parameters.Add("type", type);
			_parameters.Add("mode", mode);
			_parameters.Add("language", language);
			_parameters.Add("andOperator", andOperator);
			_parameters.Add("start", start);
			_parameters.Add("end", end);
			_parameters.Add("orderByComparator", orderByComparator);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/search", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<IDictionary<string, object>> SearchAsync(long companyId, IEnumerable<long> groupIds, IEnumerable<long> classNameIds, IEnumerable<long> classPKs, string name, string description, string type, string mode, string language, bool andOperator, int start, int end, IDictionary<string, object> orderByComparator)
		{
			var _parameters = new JsonObject();

			_parameters.Add("companyId", companyId);
			_parameters.Add("groupIds", groupIds);
			_parameters.Add("classNameIds", classNameIds);
			_parameters.Add("classPKs", classPKs);
			_parameters.Add("name", name);
			_parameters.Add("description", description);
			_parameters.Add("type", type);
			_parameters.Add("mode", mode);
			_parameters.Add("language", language);
			_parameters.Add("andOperator", andOperator);
			_parameters.Add("start", start);
			_parameters.Add("end", end);
			_parameters.Add("orderByComparator", orderByComparator);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/search", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}

		public async Task<long> SearchCountAsync(long companyId, long groupId, long classNameId, long classPK, string name, string description, string type, string mode, string language, bool andOperator)
		{
			var _parameters = new JsonObject();

			_parameters.Add("companyId", companyId);
			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);
			_parameters.Add("classPK", classPK);
			_parameters.Add("name", name);
			_parameters.Add("description", description);
			_parameters.Add("type", type);
			_parameters.Add("mode", mode);
			_parameters.Add("language", language);
			_parameters.Add("andOperator", andOperator);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/search-count", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (long)_obj;
		}

		public async Task<long> SearchCountAsync(long companyId, IEnumerable<long> groupIds, IEnumerable<long> classNameIds, IEnumerable<long> classPKs, string name, string description, string type, string mode, string language, bool andOperator)
		{
			var _parameters = new JsonObject();

			_parameters.Add("companyId", companyId);
			_parameters.Add("groupIds", groupIds);
			_parameters.Add("classNameIds", classNameIds);
			_parameters.Add("classPKs", classPKs);
			_parameters.Add("name", name);
			_parameters.Add("description", description);
			_parameters.Add("type", type);
			_parameters.Add("mode", mode);
			_parameters.Add("language", language);
			_parameters.Add("andOperator", andOperator);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/search-count", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (long)_obj;
		}

		public async Task<long> SearchCountAsync(long companyId, long groupId, long classNameId, long classPK, string keywords, string type, string mode)
		{
			var _parameters = new JsonObject();

			_parameters.Add("companyId", companyId);
			_parameters.Add("groupId", groupId);
			_parameters.Add("classNameId", classNameId);
			_parameters.Add("classPK", classPK);
			_parameters.Add("keywords", keywords);
			_parameters.Add("type", type);
			_parameters.Add("mode", mode);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/search-count", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (long)_obj;
		}

		public async Task<long> SearchCountAsync(long companyId, IEnumerable<long> groupIds, IEnumerable<long> classNameIds, IEnumerable<long> classPKs, string keywords, string type, string mode)
		{
			var _parameters = new JsonObject();

			_parameters.Add("companyId", companyId);
			_parameters.Add("groupIds", groupIds);
			_parameters.Add("classNameIds", classNameIds);
			_parameters.Add("classPKs", classPKs);
			_parameters.Add("keywords", keywords);
			_parameters.Add("type", type);
			_parameters.Add("mode", mode);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/search-count", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (long)_obj;
		}

		public async Task<IDictionary<string, object>> UpdateTemplateAsync(long templateId, long classPK, IDictionary<string, string> nameMap, IDictionary<string, string> descriptionMap, string type, string mode, string language, string script, bool cacheable, bool smallImage, string smallImageURL, IDictionary<string, object> smallImageFile, IDictionary<string, object> serviceContext)
		{
			var _parameters = new JsonObject();

			_parameters.Add("templateId", templateId);
			_parameters.Add("classPK", classPK);
			_parameters.Add("nameMap", nameMap);
			_parameters.Add("descriptionMap", descriptionMap);
			_parameters.Add("type", type);
			_parameters.Add("mode", mode);
			_parameters.Add("language", language);
			_parameters.Add("script", script);
			_parameters.Add("cacheable", cacheable);
			_parameters.Add("smallImage", smallImage);
			_parameters.Add("smallImageURL", smallImageURL);
			_parameters.Add("smallImageFile", smallImageFile);
			_parameters.Add("serviceContext", serviceContext);

			var _command = new JsonObject()
			{
				{ "/ddmtemplate/update-template", _parameters }
			};

			var _obj = await this.Session.InvokeAsync(_command);

			return (IDictionary<string, object>)_obj;
		}
	}
}